rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -1635,119 +1659,133 @@ private Optional<Federation> getFederationFromPublicKey(BtcECKey federatorPublic
         return Optional.empty();
     }
 
+    private void logMissingSignatures(BtcTransaction btcTx, Keccak256 rskTxHash, int neededSignatures, int federationSize) {
+        int missingSignatures = BridgeUtils.countMissingSignatures(btcContext, btcTx);
+        int signaturesCount = neededSignatures - missingSignatures;
+
+        logger.debug("Tx {} not yet fully signed. Requires {}/{} signatures but has {}",
+            rskTxHash, neededSignatures, federationSize, signaturesCount);
+    }
+
+    private void logReleaseBtc(BtcTransaction btcTx, byte[] rskTxHashSerialized) {
+        logger.info("Tx fully signed {}. Hex: {}", btcTx, Bytes.of(btcTx.bitcoinSerialize()));
+        eventLogger.logReleaseBtc(btcTx, rskTxHashSerialized);
+    }
+
     private void processSigning(
         FederationMember federatorMember,
         List<byte[]> signatures,
-        byte[] rskTxHash,
-        BtcTransaction btcTx,
-        Federation federation) throws IOException {
+        byte[] rskTxHashSerialized,
+        BtcTransaction btcTx) {
 
-        BtcECKey federatorBtcPublicKey = federatorMember.getBtcPublicKey();
         // Build input hashes for signatures
-        int numInputs = btcTx.getInputs().size();
-
-        List<Sha256Hash> sighashes = new ArrayList<>();
-        List<TransactionSignature> txSigs = new ArrayList<>();
-        for (int i = 0; i < numInputs; i++) {
-            TransactionInput txIn = btcTx.getInput(i);
-            Script inputScript = txIn.getScriptSig();
-            List<ScriptChunk> chunks = inputScript.getChunks();
-            byte[] program = chunks.get(chunks.size() - 1).data;
-            Script redeemScript = new Script(program);
-            sighashes.add(btcTx.hashForSignature(i, redeemScript, BtcTransaction.SigHash.ALL, false));
+        List<Sha256Hash> sigHashes = new ArrayList<>();
+        for (int i = 0; i < btcTx.getInputs().size(); i++) {
```

Comment:
Looks like we are iterating multiple times through the inputs, the sighashes, the signatures. Could this maybe be solved in a single loop?

For each input, get the sighash, get the signature, and add it?

Reply:
i tried to, but it breaks behaviour